core/trino-main/src/main/java/io/trino/execution/executor/dedicated/TaskEntry.java

```diff
@@ -97,62 +96,67 @@ public synchronized void destroy()
         pending.clear();
     }
 
-    public synchronized ListenableFuture<Void> addSplit(SplitRunner split, boolean intermediate)
+    public synchronized ListenableFuture<Void> enqueueLeafSplit(SplitRunner split)
     {
-        checkArgument(!destroyed, "Task already destroyed: %s", taskId);
+        SettableFuture<Void> done = SettableFuture.create();
+        pending.add(new QueuedSplit(split, done));
+        return done;
+    }
 
-        if (intermediate) {
-            return submit(split, false);
-        }
-        else {
-            QueuedSplit queuedSplit = new QueuedSplit(split, SettableFuture.create());
-            pending.add(queuedSplit);
-            scheduleMoreLeafSplits();
-            return queuedSplit.done();
+    /**
+     * @return true if a split was scheduled; false if no splits are pending
+     */
+    public synchronized boolean dequeueAndRunLeafSplit(Runnable doneCallback)
+    {
+        QueuedSplit split = pending.poll();
+        if (split == null) {
+            return false;
         }
+
+        runSplit(split.split())
+                .addListener(() -> {
+                    leafSplitDone(split);
+                    doneCallback.run();
+                }, directExecutor());
+
+        runningLeafSplits++;
+
+        return true;
     }
 
-    private ListenableFuture<Void> submit(SplitRunner split, boolean leaf)
+    private synchronized void leafSplitDone(QueuedSplit split)
+    {
+        runningLeafSplits--;
+        split.done().set(null);
```

Comment:
`checkArgument(!split.done.isDone())`?

Reply:
I not sure you want that.  The split could have been canceled or such since it is visible to the scheduler above.